File: 0778-swim-in-rising-water/0778-swim-in-rising-water.cpp
class Solution {
public:
    vector<pair<int,int>> dirs = {{1,0}, {-1,0}, {0,1}, {0,-1}};
    int swimInWater(vector<vector<int>>& grid) {
        priority_queue<vector<int>, vector<vector<int>>, greater<>> pq; // elevation, row, col
        int n = grid.size();

        vector<vector<bool>> visited(n, vector<bool>(n, false));

        pq.push({0, 0, 0});
        visited[0][0] = true;
        int res = grid[0][0];

        // pop pq
        // always pick the lowest
        // keep track of last
        while (!pq.empty()) {
            int elevation = pq.top()[0];
            int r = pq.top()[1];
            int c = pq.top()[2];
            
            pq.pop();
            res = max(res, elevation);

            if (r == n-1 && c == n-1) {
                return res;
            }
            
            for (auto& dir : dirs) {
                int newR = r + dir.first;
                int newC = c + dir.second;

                if (newR < 0 || newR >= n || newC < 0 || newC >= n) {
                    continue;
                }

                if (visited[newR][newC]) {
                    continue;
                }

                visited[newR][newC] = true;
                pq.push({grid[newR][newC], newR, newC});
            }
        }

        return -1;
    }
};

File: 0778-swim-in-rising-water/0778-swim-in-rising-water.py
class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        # visited grid to avoid revisiting
        # minHeap with elevation, r, c
        # keep popping until meet bottom right

        dirs = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        n = len(grid)
        visited = [[False] * n  for _ in range(n)]
        minHeap = [[grid[0][0], 0, 0]]

        while minHeap:
            height, r, c = heapq.heappop(minHeap)

            if r == n-1 and c == n-1:
                return height

            if visited[r][c]:
                continue

            visited[r][c] = True

            for dir in dirs:
                newR = r + dir[0]
                newC = c + dir[1]

                if (newR < 0 or newR >= n or newC < 0 or newC >= n):
                    continue

                newHeight = max(height, grid[newR][newC])
                heapq.heappush(minHeap, [newHeight, newR, newC])


        return -1
        